zh: docs/ops/integrations/spire/ (#13872)
istio/istio.io: content/zh/docs/ops/integrations/spire/index.md

---
title: SPIRE
description: 如何配置 Istio 以集成 SPIRE 通过 Envoy 的 SDS API 获取加密身份。
weight: 31
keywords: [kubernetes,spiffe,spire]
aliases:
owner: istio/wg-networking-maintainers
test: yes
---

> [SPIRE](https://spiffe.io/docs/latest/spire-about/spire-concepts/) 是 SPIFFE 规范的一个可用于生产环境的实现，它执行节点和工作负载的认证，以便在异构环境中安全地为运行的工作负载颁发加密身份。通过与 [Envoy 的 SDS API](https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret) 集成，可以将 SPIRE 配置为 Istio 工作负载的加密身份源。Istio 可以检测到实现了 Envoy SDS API 的 UNIX 域套接字的存在，并允许 Envoy 直接从其中通信并获取身份。

Reviewer: Unix domian socket，感觉是不是不翻译更好点？

Author: 我觉得有通用翻译的名词可以翻译。

> 与默认的 Istio 身份管理相比，与 SPIRE 的集成提供了灵活的认证选项。例如，SPIRE 的插件架构使得可以选择 Kubernetes 命名空间和服务账户认证之外的多样化工作负载认证选项。SPIRE 的节点认证将认证扩展到工作负载运行的物理或虚拟硬件上。

Reviewer 2: ```suggestion
与默认的 Istio 身份管理相比，与 SPIRE 的集成提供了灵活的认证选项。例如，SPIRE 的插件架构使您可以选择 Kubernetes
命名空间和服务账户认证之外的多样化工作负载认证选项。SPIRE 的节点认证将认证扩展到工作负载所运行的物理或虚拟硬件上。
```
长句可以适当折行

> 要了解关于如何将 SPIRE 与 Istio 集成的快速演示，请参阅 [通过 Envoy 的 SDS API 将 SPIRE 集成为 CA]({{< github_tree >}}/samples/security/spire)。

Reviewer 2: ```suggestion
要了解关于如何将 SPIRE 与 Istio 集成的快速演示，请参阅[通过 Envoy 的 SDS API 将 SPIRE 集成为 CA]({{< github_tree >}}/samples/security/spire)。
```

{{< warning >}}

> 请注意，此集成需要 `istioctl` 和数据平面均为版本 1.14+。

Reviewer 2: ```suggestion
请注意，此集成需要 `istioctl` 和数据平面均为 1.14 及更高版本。
```

{{< /warning >}}

此集成与 Istio 升级兼容。

## 安装 SPIRE {#install-spire}

### 选项 1：快速开始 {#option-1-quick-start}

Istio 提供了一个基本的示例安装，可以快速启动 SPIRE：

{{< text syntax=bash snip_id=install_spire_with_controller_manager >}}
$ kubectl apply -f @samples/security/spire/spire-quickstart.yaml@
{{< /text >}}

> 这将在您的集群中部署 SPIRE，以及两个附加组件：[SPIFFE CSI 驱动](https://github.com/spiffe/spiffe-csi)——用于与节点中的其他 pod 共享 SPIRE 代理的 UNIX 域套接字 — 以及 [SPIRE Controller Manager](https://github.com/spiffe/spire-controller-manager)，它负责进行工作负载注册并在 Kubernetes 中建立联合关系。请参阅 [安装 Istio](#install-istio) 以配置 Istio 并集成 SPIFFE CSI 驱动。

Reviewer 2: ```suggestion
这将在您的集群中部署 SPIRE，以及下述两个附加组件：

- [SPIFFE CSI 驱动](https://github.com/spiffe/spiffe-csi)，用于与节点中的其他 Pod 共享 SPIRE 代理的 UNIX 域套接字；
- [SPIRE Controller Manager](https://github.com/spiffe/spire-controller-manager)，它负责注册工作负载并在
  Kubernetes 中建立联合关系。请参阅[安装 Istio](#install-istio) 以配置 Istio 并集成 SPIFFE CSI 驱动。
```
可以稍微改一下结构，提高可读性

### 选项 2：配置自定义 SPIRE 安装 {#option-2-configure-a-custom-spire-installation}

> 请参阅 [SPIRE 的 Kubernetes 快速入门指南](https://spiffe.io/docs/latest/try/getting-started-k8s/) 以开始将 SPIRE 部署到您的 Kubernetes 环境中。请参阅 [SPIRE CA 集成先决条件](#spire-ca-integration-prerequisites) 以获取有关如何配置 SPIRE 以与 Istio 部署集成的更多信息。

Reviewer 2: ```suggestion
请参阅 [SPIRE 的 Kubernetes 快速入门指南](https://spiffe.io/docs/latest/try/getting-started-k8s/)
以开始将 SPIRE 部署到您的 Kubernetes 环境中。请参阅 [SPIRE CA 集成先决条件](#spire-ca-integration-prerequisites)
以获取有关如何配置 SPIRE 以与 Istio 部署集成的更多信息。
```

#### SPIRE CA 集成先决条件 {#spire-ca-integration-prerequisites}

要将您的 SPIRE 部署与 Istio 集成，需要配置 SPIRE：

1. 访问 [SPIRE 代理参考](https://spiffe.io/docs/latest/deploying/spire_agent/#agent-configuration-file) 并将 SPIRE 代理套接字路径配置为与 Envoy SDS 定义的套接字路径匹配。

    {{< text plain >}}
    socket_path = "/run/secrets/workload-spiffe-uds/socket"
    {{< /text >}}

> 1. 通过部署 [SPIFFE CSI 驱动](https://github.com/spiffe/spiffe-csi) 与节点内的 pod 共享 SPIRE 代理套接字。驱动程序的 `-workload-api-socket-dir` 参数应为套接字目录的挂载位置。

Reviewer 3: ```suggestion
1. 通过部署 [SPIFFE CSI 驱动](https://github.com/spiffe/spiffe-csi) 与节点内的 Pod 共享 SPIRE 代理套接字，通过设置驱动程序的 `-workload-api-socket-dir` 参数来指定套接字目录的挂载位置。
```

请参阅 [安装 Istio](#install-istio) 以配置 Istio 以集成 SPIFFE CSI 驱动。

{{< tip >}}
如果在 Istio 代理启动之前 SPIRE 尚未创建套接字，则 Istio 将成为 Envoy SDS 监听器。此时的时间由自定义 IstioOperator 控制。
{{< /tip >}}

## 安装 Istio {#install-istio}

### 选项  1：使用 SPIRE Controller Manager 配置工作负载注册 {#option-1-configuration-for-workload-registration-with-the-spire-controller-manager}

> 通过部署 [SPIRE Controller Manager](https://github.com/spiffe/spire-controller-manager) 和 SPIRE 服务器，可以自动为与 [ClusterSPIFFEID](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) 自定义资源中定义的选择器匹配的每个新 pod 进行注册。

Reviewer 3: ```suggestion
通过部署 [SPIRE Controller Manager](https://github.com/spiffe/spire-controller-manager) 和 SPIRE 服务器，可以自动为与 [ClusterSPIFFEID](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) 自定义资源中定义的选择器匹配的每个新 Pod 进行注册。
```

> 必须在安装 Istio 之前应用 ClusterSPIFFEID。此外，必须配置 Ingress-gateway pod 以匹配 ClusterSPIFFEID 中定义的选择器。如果在安装期间未自动创建 Ingress Gateway 工作负载的注册条目，则工作负载将无法达到 `Ready` 状态。

Reviewer 2: ```suggestion
必须在安装 Istio 之前应用 ClusterSPIFFEID。此外，必须配置 Ingress-gateway Pod 以匹配 ClusterSPIFFEID 中定义的选择器。
如果在安装期间未自动创建 Ingress Gateway 工作负载的注册条目，则工作负载将无法达到 `Ready` 状态。
```

1. 创建 ClusterSPIFFEID 示例：

    {{< text syntax=bash snip_id=create_clusterspiffeid >}}
    $ kubectl apply -f - <<EOF
    apiVersion: spire.spiffe.io/v1alpha1
    kind: ClusterSPIFFEID
    metadata:
      name: example
    spec:
      spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
      podSelector:
        matchLabels:
          spiffe.io/spire-managed-identity: "true"
    EOF
    {{< /text >}}

>     示例 ClusterSPIFFEID 允许自动注册所有具有 `spiffe.io/spire-managed-identity: "true"` 标签的工作负载。对于具有此标签的 pod，将提取 `spiffeIDTemplate` 中指定的值以形成 SPIFFE ID。

Reviewer 3: ```suggestion
    示例 ClusterSPIFFEID 允许自动注册所有具有 `spiffe.io/spire-managed-identity: "true"` 标签的工作负载。对于具有此标签的 Pod，将提取 `spiffeIDTemplate` 中指定的值以形成 SPIFFE ID。
```

                      value:
                        - name: wait-for-spire-socket
                          image: busybox:1.28
                          volumeMounts:
                            - name: workload-socket
                              mountPath: /run/secrets/workload-spiffe-uds
                              readOnly: true
                          env:
                            - name: CHECK_FILE
                              value: /run/secrets/workload-spiffe-uds/socket
                          command:
                            - sh
                            - "-c"
                            - |-
                              echo "$(date -Iseconds)" Waiting for: ${CHECK_FILE}
                              while [[ ! -e ${CHECK_FILE} ]] ; do
                                echo "$(date -Iseconds)" File does not exist: ${CHECK_FILE}
                                sleep 15
                              done
                              ls -l ${CHECK_FILE}
    EOF
    {{< /text >}}

1. 应用配置：

    {{< text syntax=bash snip_id=apply_istio_operator_configuration >}}
    $ istioctl install --skip-confirmation -f ./istio.yaml
    {{< /text >}}

> 1. 检查 Ingress-gateway pod 状态：

Reviewer 3: ```suggestion
1. 检查 Ingress-gateway Pod 状态：
```

    {{< text syntax=bash snip_id=none >}}
    $ kubectl get pods -n istio-system
    NAME                                    READY   STATUS    RESTARTS   AGE
    istio-ingressgateway-5b45864fd4-lgrxs   1/1     Running   0          17s
    istiod-989f54d9c-sg7sn                  1/1     Running   0          23s
    {{< /text >}}

    由于与之对应的注册条目已在 SPIRE 服务器上自动创建，Ingress-gateway pod 状态为 `Ready`。Envoy 能够从 SPIRE 获取加密身份。

请注意，在[快速入门](#option-1-quick-start) 部分中使用了 `SPIRE Controller Manager`。

### 选项 2：手动工作负载注册 {#option-2-configuration-for-manual-workload-registration-with-spire}

1. [下载 Istio](/zh/docs/setup/getting-started/#download)。

> 1. 在[部署 SPIRE](#install-spire)到您的环境并确保所有部署都处于 `Ready` 状态后，配置  Istio，包括对 Ingress-gateway 和 istio-proxy 的自定义补丁。

Reviewer: ```suggestion
1. 在[部署 SPIRE](#install-spire)到您的环境并确保所有部署都处于 `Ready` 状态后，配置 Istio ，包括对 Ingress-gateway 和 istio-proxy 的自定义补丁。
```

                          command:
                            - sh
                            - "-c"
                            - |-
                              echo "$(date -Iseconds)" Waiting for: ${CHECK_FILE}
                              while [[ ! -e ${CHECK_FILE} ]] ; do
                                echo "$(date -Iseconds)" File does not exist: ${CHECK_FILE}
                                sleep 15
                              done
                              ls -l ${CHECK_FILE}
    EOF
    {{< /text >}}

1. 应用配置：

    {{< text syntax=bash snip_id=none >}}
    $ istioctl install --skip-confirmation -f ./istio.yaml
    {{< /text >}}

1. 检查 Ingress-gateway pod 状态：

    {{< text syntax=bash snip_id=none >}}
    $ kubectl get pods -n istio-system
    NAME                                    READY   STATUS    RESTARTS   AGE
    istio-ingressgateway-5b45864fd4-lgrxs   0/1     Running   0          20s
    istiod-989f54d9c-sg7sn                  1/1     Running   0          25s
    {{< /text >}}

    只有在 SPIRE 服务器上为 Ingress-gateway pod 和数据平面容器创建了相应的注册条目后，它们才会达到 `Ready` 状态。
    接着，Envoy 将能够从 SPIRE 中获取加密身份信息。
    请参阅[注册工作负载](#register-workloads)以为您的网格中的服务注册条目。

> Istio配置与入口网关以及将要注入到工作负载Pod上的Sidecar共享`spiffe-csi-driver`，从而使它们能够访问SPIRE代理的UNIX域套接字。

Reviewer 2: ```suggestion
Istio 配置与入口网关以及将要注入到工作负载 Pod 上的 Sidecar 共享 `spiffe-csi-driver`，从而使它们能够访问 SPIRE 代理的 UNIX 域套接字。
```

> 此配置还向网关添加了一个initContainer，它将等待SPIRE创建UNIX域套接字，然后再启动istio-proxy。如果SPIRE代理尚未准备好，或者尚未使用相同的套接字路径正确配置，Ingress网关的initContainer将永远等待。

Reviewer 2: ```suggestion
此配置还向网关添加了一个 initContainer，它将等待 SPIRE 创建 UNIX 域套接字，然后再启动 istio-proxy。
如果 SPIRE 代理尚未准备好，或者尚未使用相同的套接字路径正确配置，Ingress 网关的 initContainer 将永远处于等待状态。
```

## 注册工作负载 {#register-workloads}

本节描述了在SPIRE服务器中注册工作负载的可用选项。

### 选项1：使用 SPIRE 控制器管理器进行注册 {#option-1-registration-using-the-spire-controller-manager}

> 将为与[ClusterSPIFFEID](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md)自定义资源中定义的选择器匹配的每个新 Pod 自动注册新条目。请参阅[使用 SPIRE 控制器管理器进行工作负载注册的配置](#option-1-configuration-for-workload-registration-with-the-spire-controller-manager)以获取示例ClusterSPIFFEID配置。

Reviewer 2: ```suggestion
将为与 [ClusterSPIFFEID](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md)自定义资源中定义的选择器匹配的每个新 Pod 自动注册新条目。
请参阅[使用 SPIRE 控制器管理器进行工作负载注册的配置](#option-1-configuration-for-workload-registration-with-the-spire-controller-manager)以获取示例 ClusterSPIFFEID 配置。
```

1. 部署示例工作负载：

    {{< text syntax=bash snip_id=apply_sleep >}}
    $ istioctl kube-inject --filename @samples/security/spire/sleep-spire.yaml@ | kubectl apply -f -
    {{< /text >}}

>     除了需要 `spiffe.io/spire-managed-identity` 标签之外，工作负载还需要使用 SPIFFE CSI 驱动器卷来访问 SPIRE 代理套接字。为了实现这一点，您可以利用[安装 Istio](#install-istio) 部分中的`spire` pod 注解模板，或者将 CSI 卷添加到您的工作负载的部署规范中。这两种方法都在下面的示例片段中进行了突出显示：

Reviewer 3: ```suggestion
    除了需要 `spiffe.io/spire-managed-identity` 标签之外，工作负载还需要使用 SPIFFE CSI 驱动器卷来访问 SPIRE 代理套接字。为了实现这一点，您可以利用[安装 Istio](#install-istio) 部分中的 `spire` Pod 注解模板，或者将 CSI 卷添加到您的工作负载的部署规范中。这两种方法都在下面的示例片段中进行了突出显示：
```

          matchLabels:
            app: sleep
      template:
          metadata:
            labels:
              app: sleep
              spiffe.io/spire-managed-identity: "true"
            # Injects custom sidecar template
            annotations:
                inject.istio.io/templates: "sidecar,spire"
          spec:
            terminationGracePeriodSeconds: 0
            serviceAccountName: sleep
            containers:
            - name: sleep
              image: curlimages/curl
              command: ["/bin/sleep", "3650d"]
              imagePullPolicy: IfNotPresent
              volumeMounts:
                - name: tmp
                  mountPath: /tmp
              securityContext:
                runAsUser: 1000
            volumes:
              - name: tmp
                emptyDir: {}
              # CSI volume
              - name: workload-socket
                csi:
                  driver: "csi.spiffe.io"
                  readOnly: true
    {{< /text >}}

请参阅[验证为工作负载创建的身份](#verifying-that-identities-were-created-for-workloads)
以检查已颁发的身份。

请注意，在[快速入门](#option-1-quick-start)部分中使用了`SPIRE控制器管理器`。

### 选项 2：手动注册 {#option-2-manual-registration}

> 为提高工作负载的认证安全性和稳健性，SPIRE 能够根据不同参数的一组选择器值来进行验证。如果您是通过按照[快速入门](#option-1-quick-start)中的步骤来安装 `SPIRE`，则可以跳过这些步骤，因为它使用自动注册。

Reviewer 2: ```suggestion
为提高工作负载的认证安全性和稳健性，SPIRE 能够根据不同参数的一组选择器值来进行验证。
如果您是参照[快速入门](#option-1-quick-start)中的步骤来安装 `SPIRE`，则可以跳过这些步骤，
因为 SPIRE 使用自动注册的机制。
```

> 1. 使用一组选择器为 Ingress Gateway 生成条目，例如 pod 名称和 pod UID：

Reviewer 2: ```suggestion
1. 使用一组选择器为 Ingress Gateway 生成条目，例如 Pod 名称和 Pod UID：
```
译文中的 Pod 通常首字母大写会更好看一些，也是为了保持所有页面一致的风格。

| `default_svid_name`        | 用于 Envoy SDS 的默认 `X509-SVID` 的 TLS 证书资源名称   | 默认          |
| `default_bundle_name`      | 用于 Envoy SDS 的默认 X.509 bundle 的验证上下文资源名称 | 空             |
| `default_all_bundles_name` | 用于 Envoy SDS 的所有 bundle（包括联邦的）的验证上下文资源名称 | ROOTCA       |
这将允许 Envoy 直接从 SPIRE 获取联邦 bundle。

### 创建联邦注册条目 {#create-federated-registration-entries}

* 如果使用 SPIRE Controller Manager，请通过将 [ClusterSPIFFEID CR](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) 的 `federatesWith` 字段设置为您希望 pod 与之联邦的信任域来为工作负载创建联邦条目：

    {{< text syntax=yaml snip_id=none >}}
    apiVersion: spire.spiffe.io/v1alpha1
    kind: ClusterSPIFFEID
    metadata:
      name: federation
    spec:
      spiffeIDTemplate: "spiffe://{{ .TrustDomain }}/ns/{{ .PodMeta.Namespace }}/sa/{{ .PodSpec.ServiceAccountName }}"
      podSelector:
        matchLabels:
          spiffe.io/spire-managed-identity: "true"
      federatesWith: ["example.io", "example.ai"]
    {{< /text >}}

* 要进行手动注册，请参阅[为联邦创建注册条目](https://spiffe.io/docs/latest/architecture/federation/readme/#create-registration-entries-for-federation)。

## 清理 SPIRE {#cleanup-spire}

> 如果您是通过Istio提供的快速入门SPIRE部署安装的SPIRE，请使用以下命令来删除这些Kubernetes资源：

Reviewer 3: ```suggestion
如果您是通过 Istio 提供的快速入门 SPIRE 部署安装的 SPIRE，请使用以下命令来删除这些 Kubernetes 资源：
```

1. 检查证书并确认 SPIRE 是颁发者：

    {{< text syntax=bash snip_id=get_svid_subject >}}
    $ openssl x509 -in chain.pem -text | grep SPIRE
        Subject: C = US, O = SPIRE, CN = sleep-5f4d47c948-njvpk
    {{< /text >}}

## SPIFFE 联邦 {#spiffe-federation}

SPIRE 服务器能够对来自不同信任域的 SPIFFE 身份进行认证，这被称为 SPIFFE 联邦。

可以配置 SPIRE 代理通过 Envoy SDS API 向 Envoy 推送联邦捆绑包，从而使 Envoy 能够使用[验证上下文](https://spiffe.io/docs/latest/microservices/envoy/#validation-context)来验证对等证书并信任来自另一个信任域的工作负载。
要使 Istio 能够通过 SPIRE 集成实现 SPIFFE 身份联合，请参阅 [SPIRE 代理 SDS 配置](https://github.com/spiffe/spire/blob/main/doc/spire_agent.md#sds-configuration)，并为您的 SPIRE 代理配置文件设置以下 SDS 配置值。

| 配置                       | 描述                                                | 资源名称    |
| -------------------------- | --------------------------------------------------- | ------------- |
| `default_svid_name`        | 用于 Envoy SDS 的默认 `X509-SVID` 的 TLS 证书资源名称   | 默认          |
| `default_bundle_name`      | 用于 Envoy SDS 的默认 X.509 bundle 的验证上下文资源名称 | 空             |
| `default_all_bundles_name` | 用于 Envoy SDS 的所有 bundle（包括联邦的）的验证上下文资源名称 | ROOTCA       |
这将允许 Envoy 直接从 SPIRE 获取联邦 bundle。

### 创建联邦注册条目 {#create-federated-registration-entries}

> * 如果使用 SPIRE Controller Manager，请通过将 [ClusterSPIFFEID CR](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) 的 `federatesWith` 字段设置为您希望 pod 与之联邦的信任域来为工作负载创建联邦条目：

Reviewer 2: ```suggestion
- 如果使用 SPIRE Controller Manager，请通过将 [ClusterSPIFFEID CR](https://github.com/spiffe/spire-controller-manager/blob/main/docs/clusterspiffeid-crd.md) 的 `federatesWith` 字段设置为您希望 pod 与之联邦的信任域来为工作负载创建联邦条目：
```
lint 可能是说上文用 `-`，这里用 `*` 不一致吗，奇怪英文怎么 pass 的
